feat: holistic SAFE Txids handling

gomesalexandre opened the pull request:
## Description

Follows-up on https://github.com/shapeshift/web/pull/7506 by gracefully handling SAFE Tx hashes / Txids across the app.

Actually does - cherry-picks https://github.com/shapeshift/web/pull/7698 + adds SAFE handling on top of that. ~~Note: This does *not* handle sends, as they will require a new status step.~~

Tested in:

- [x] RFOX
  - [x] Stake
  - [x] Unstake
  - [ ] Claim (unable to test with SAFE)
  - [x] Change addy
- [x] THORChain LP - correctly disabled ✅ - we cannot do this because of `tx.origin` shenanigans
- [x] DeFi section
  - [x] Savers Deposit is disabled ✅ - we cannot do this because of shenanigans re: `tx.origin`
  - [x] ETH/FOX LP
    - [x] Withdraw
    - [x] Deposit
  - [x] ETH/FOX Staking
    - [x] Withdraw
    - [x] Deposit
    - [x] Claim
- [x] FOXy
    - [x] Withdraw
    - [x] Deposit - implemented for the sake of consistency - deposits are disabled for FOXy
    - [x] Claim (but untestable currently, gotta wait 13 days)
- [x] ~~THORChain Lending~~ effectively ded
- [x] Sends
- [ ] Swapper
  - [x] Regression test swapper and ensure SAFE Tx detection is still happy
  - [x] RFOX Arbitrum Bridge - unapplicable
  - [x] Arb Claims (implemented but untestable)

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

<!-----------------------------------------------------------------------------
If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.
------------------------------------------------------------------------------>

closes N/A

## Risk
> High Risk PRs Require 2 approvals

<!-----------------------------------------------------------------------------
Outline the scope of your changes and the risk associated with them. You must use your discretion as an engineer to determine the potential impact of your changes.

WARNING: If your PR introduces a new on-chain transaction or modifies an existing one, please add the "High-Risk" label to this PR and ask for 2 reviewers to approve it before merging.

E.g. an upgrade to `hdwallet` or core state management would be considered higher risk, and might require a full regression test. UI or isolated view changes, or something behind a feature flag may have near zero risk. Small bug fixes might require testing isolated to the specific fix.
------------------------------------------------------------------------------>

> What protocols, transaction types or contract interactions might be affected by this PR?

High - Tx status detection could be borked for both SAFE Txs in swapper, and SAFE and non-safe Txs across the app

## Testing

<!-----------------------------------------------------------------------------
We treat every PR to be merged with the same scrutiny as if we were merging directly to production.

Your PR will not be merged if you do not complete the sections below for our engineering and operations teams to test.
------------------------------------------------------------------------------>

- Do a full regression test of all the above, non-SAFE Txs status detection and their Tx link should still be happy (including approvals)
- Ensure SAFE Txs are properly handled, with the "x out of y signed" copy/polling and correct Tx links (SAFE Tx link when queued, on-chain Tx afterwards)

### Engineering

<!-----------------------------------------------------------------------------
Include sufficient information here for an engineer to test your PR. This may include how to test locally, in a built environment, changes to infrastructure etc.
------------------------------------------------------------------------------>

### Operations

<!-----------------------------------------------------------------------------
If your changes have a user-facing impact, describe how a non-technical QA team can functionally test your changes in a preview environment.

If they are not user-facing please describe how to test for any regressions that may occur.
------------------------------------------------------------------------------>

## Screenshots (if applicable)

### RFOX

- Stake: https://jam.dev/c/836607d0-9010-4e50-8ee9-4e2fd774418f
- Unstake: https://jam.dev/c/8a0a0c16-b5df-4c9a-b979-58dc0ddd6dbd
- Claim: Unable to test
- Change addy: https://jam.dev/c/0cedf9e1-c3e2-4da6-9b5f-2d4dfd1b9f1c

### ETH/FOX LP

- Withdraw: https://jam.dev/c/011e08c7-d274-4b89-a3b7-8df2760b6d5d
- Deposit: https://jam.dev/c/a8f911eb-2c59-4107-b53c-5f2f08a37480


### ETH/FOX Staking

- Deposit: https://jam.dev/c/be48cc4d-496f-43e8-9c47-fccc3fb694b9
- Withdraw: https://jam.dev/c/c5881ada-0ddf-4616-a900-48762196690a
- Claim: https://jam.dev/c/d0ded2e5-35ec-43f4-a62a-38f4ff1380c1

### FOXy

- Deposit: N/A, disabled
- Delayed Withdraw: https://jam.dev/c/0e06ea66-5c47-401a-bac4-bbf65b6ff872
- Instant Withdraw: https://jam.dev/c/c7b043cc-ea2a-46c1-95a6-65393c3218d5
- Claim: Untestable in sc currently

### Sends

https://jam.dev/c/9f25c4ee-c2f3-45c9-a17e-2c498f55d21e

### Swapper Regression Test

https://jam.dev/c/43dc3493-7bc9-4af0-9345-d46ca8a8bb36

<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->

gomesalexandre commented on src/hooks/queries/useSafeTx.ts:
```diff
@@ -1,22 +1,29 @@
 
 type UseSafeTxQueryArgs = {
-  chainId: ChainId
+  accountId: AccountId | undefined
```
The guts of this PR and its plumbing: we now pass an `AccountId` to `useSafeTxQuery` and consume SAFE TX data holistically.

The reason why we do said plumbing is so that we can do preliminary checks in `fetchSafeTransactionInfo` re: is the current address a SAFE account (with an infinitely cached query, passing the curried `queryFn` leveraging web provider up-top so we don't overfetch), and early bail if so, so that the app doesn't spew requests to SAFE API that we don't need.

gomesalexandre commented on packages/swapper/src/safe-utils/index.ts:
```diff
@@ -10,10 +11,26 @@ import type {
+  }
+  const chainId = fromAccountId(accountId).chainId
   if (!isEvmChainId(chainId)) return { transaction: null, isSafeTxHash: false }
+  const isSmartContractAddress = await fetchIsSmartContractAddressQuery(
+    fromAccountId(accountId).account,
+    chainId,
+  )
+
+  if (!isSmartContractAddress) {
+    // Assume any smart contract address is a SAFE, and by extension, any non-smart contract address is not
+    // This is super naive, but this is exactly the same way SAFE does it
+    // https://github.com/safe-global/safe-core-sdk/blob/ea0d5018a93f294dfd891e6c8963edcb96431876/packages/protocol-kit/src/Safe.ts#L303
+    return { transaction: null, isSafeTxHash: false }
+  }
```
Also the guts of this PR, avoids spamming SAFE API, https://github.com/shapeshift/web/pull/7525/files#r1711302800

gomesalexandre commented on src/pages/RFOX/components/Claim/ClaimStatus.tsx:
```diff
@@ -91,9 +95,20 @@ export const ClaimStatus: React.FC<Pick<ClaimRouteProps, 'headerComponent'> & Cl
-    () => getTxLink({ txId, defaultExplorerBaseUrl: claimAsset?.explorerTxLink ?? '' }),
-    [claimAsset?.explorerTxLink, txId],
+    () =>
+      getTxLink({
```
Don't forget to remove TODO of `getTxLink` - see https://github.com/shapeshift/web/pull/7506#discussion_r1725550786

gomesalexandre commented on src/features/defi/providers/fox-farming/components/FoxFarmingManager/Claim/ClaimStatus.tsx:
```diff
@@ -156,6 +188,53 @@ export const ClaimStatus: React.FC<ClaimStatusProps> = ({ accountId }) => {
+        txId: maybeSafeTx.transaction.transactionHash,
+        defaultExplorerBaseUrl: feeAsset.explorerTxLink,
+        accountId,
+        // on-chain Tx
```
So.. thinking about this some more, this seems like a lot of flaky and duplicated logic - if *isSafeTxHash* we should pass both a `safeTxHash` (defined) and `txid` (maybe undefined)

gomesalexandre commented:
Opening for early review - note THORChain LP hasn't been tested given the current hardfork shenanigans and may require some touchups before this goes in, but everything non-THOR should be testable.

NeOMakinG reviewed (commented):
A few suggestions, questions or nitpicks

Looks solid overall from a code point of view, pending runtime testing which will also be very big!

gomesalexandre commented on src/features/defi/providers/fox-farming/components/FoxFarmingManager/Claim/ClaimStatus.tsx:
```diff
@@ -128,20 +137,43 @@ export const ClaimStatus: React.FC<ClaimStatusProps> = ({ accountId }) => {
   const confirmedTransaction = useAppSelector(gs => selectTxById(gs, serializedTxIndex))
 
+  const status = useMemo(() => {
+    if (maybeSafeTx?.isQueuedSafeTx) {
+      return TxStatus.Pending
+    }
+
+    if (maybeSafeTx?.isExecutedSafeTx) {
+      return TxStatus.Confirmed
+    }
+
```
Gut feel is having SAFE data for *all* Txs vs. lazily may not be a good idea, though we early bail on sc checks so it may or may not be safe, will take a look!
Conceptually like the idea minus the eager vs. lazy ramifications.

NeOMakinG replied:
Regarding the amount of work, we shouldn't include this in this PR anyway and could be done later in a follow up if we choose

gomesalexandre commented on src/features/defi/providers/fox-farming/components/FoxFarmingManager/Claim/ClaimStatus.tsx:
```diff
@@ -128,20 +137,43 @@ export const ClaimStatus: React.FC<ClaimStatusProps> = ({ accountId }) => {
-    if (confirmedTransaction && confirmedTransaction.status !== 'Pending') {
+    if (status && status !== TxStatus.Pending) {
+      const usedGasFeeCryptoPrecision = (() => {
+        if (maybeSafeTx?.transaction?.gasUsed)
```
Would prefer to keep the SAFE types unmassaged here, don't see any reason why we wouldn't

gomesalexandre commented on src/hooks/useTxDetails/useTxDetails.ts:
```diff
@@ -110,6 +112,13 @@ export const useTxDetails = (txId: string): TxDetails => {
   // This should never happen, but if it does we should not continue
   if (!tx) throw Error('Transaction not found')
 
+  const accountId = useMemo(() => deserializeTxIndex(txId).accountId, [txId])
+
+  const { data: maybeSafeTx } = useSafeTxQuery({
+    maybeSafeTxHash: txId,
+    accountId,
+  })
```
See https://github.com/shapeshift/web/pull/7525#discussion_r1752209937 for the rationale, though once again I like the idea - if we can do this lazily enough, I'm all up for it.

gomesalexandre replied:
Summoning @kaladinlight too, he may already be sad at the idea of having maybe SAFE data in `Tx`

NeOMakinG replied:
I'm pretty agnostic anyway, happy with both but I'm asking questions to make sure we thought about everything!

kaladinlight replied:
The safe tx state should not really be a persistent or tracked thing in general I feel like. I don't even think this hook here is relevant as we should not be adding a tx to the txHistory slice unless it is on chain. Unless there is a good agreed upon reason to build out more safe info tracking within standard tx history type state, I would say we should only be doing safe things where necessary (tx status state steppers)

gomesalexandre commented on src/features/defi/providers/foxy/components/FoxyManager/Deposit/components/Status.tsx:
```diff
@@ -51,25 +87,63 @@ export const Status = () => {
+      accountId,
+      maybeSafeTx,
+    })
+  }, [accountId, feeAsset, maybeSafeTx, state?.txid])
```
Not sure about this one, hook vs. function call seems pretty similar to me

NeOMakinG replied:
You're right, it doesn't change anything except those little dependencies

NeOMakinG reviewed (commented):
Single Signer
- [x] RFOX
  - [x] Stake (Lost the jam link)
  - [x] Unstake https://jam.dev/c/3cdd16fe-8b40-4703-8b3f-44e9636b6ca4
  - [ ] Claim (unable to test with SAFE, I don't have any claims yet)
  - [x] Change addy
- [ ] THORChain LP - Smart contract wallet not supported
- [x] DeFi section
  - [x] Savers Deposit is disabled ✅ - we cannot do this because of shenanigans re: `tx.origin`
  - [x] ETH/FOX LP 
    - [x] Withdraw same as deposit
    - [x] Deposit ⁉️ (https://jam.dev/c/7a96495b-44a6-498a-a652-6b8febca743c) Deposit went successful but has never been confirmed in the UI
  - [ ] ETH/FOX Staking
    - [x] Withdraw It has never been confirmed in the UI but it worked
    - [x] Deposit https://jam.dev/c/2115d324-b9fa-4080-abdb-5a47999b6dae
    - [ ] Claim - I couldn't claim, the UI crashes https://jam.dev/c/3a869870-ce80-44fe-a3e3-ba1340b8bb4c
- [ ] FOXy (Couldn't test, I don't have any foxy opportunities)
    - [ ] Withdraw
    - [ ] Deposit
    - [ ] Claim
- [x] Sends https://jam.dev/c/93b0b2d5-e47f-4f68-9a13-1d3a7ddf33f2
- [x] Swapper
  - [x] Regression test swapper and ensure SAFE Tx detection is still happy (https://jam.dev/c/e19550bb-7572-4e7b-a4d2-88ff57b506e5)
  - [ ] RFOX Arbitrum Bridge - unapplicable
  - [ ] Arb Claims (implemented but untestable)


## Feedback
There is a weird thing with the current reward address
  
![image](https://github.com/user-attachments/assets/a7dba7a8-5c13-415d-a5bb-ad532dbd408d)

It doesn't display any address for some reason

Deposit into ETH/FOX farming
![image](https://github.com/user-attachments/assets/c444a249-50c5-45c2-b51f-bbc64c189ff5)
![image](https://github.com/user-attachments/assets/d2f49d91-eb93-4ef7-a450-c19f6fd52805)

The UI deserves a little bit of padding, same for withdraw and maybe same for claim

### Claiming
![image](https://github.com/user-attachments/assets/a5c2cc47-54f5-4819-82a5-2ca3d23dad76)

The header is going out of the container (look top left and top right)

NeOMakinG commented on src/pages/RFOX/components/Claim/ClaimStatus.tsx:
```diff
@@ -51,9 +58,36 @@ export const ClaimStatus: React.FC<Pick<ClaimRouteProps, 'headerComponent'> & Cl
+        ],
+        body: [
+          'RFOX.claimPending',
+          { amount: bnOrZero(claimAmountCryptoPrecision).toFixed(8), symbol: claimAsset.symbol },
```
![image](https://github.com/user-attachments/assets/a4799b38-e0a8-41b6-b5a6-bb0d17656299)

All those amounts from the status page are displaying unneeded zeros (0.06250000 FOX instead of 0.0625 FOX), you could avoid this by using `toCrypto` with omitDecimalTrailingZeros

gomesalexandre commented:
> * [x]  Deposit ⁉️ ([jam.dev/c/7a96495b-44a6-498a-a652-6b8febca743c](https://jam.dev/c/7a96495b-44a6-498a-a652-6b8febca743c)) Deposit went successful but has never been confirmed in the UI

I can see this isn't a SAFE Tx (or at least doesn't seems to be from what I can see in your Jam) - tested both with SAFE and native, and was able to repro with the latter. Applies to FOX farming as well. Fixed in b75a3ce48e:

### UNI-V2 Deposit:

- SAFE: https://jam.dev/c/7f69c31d-7dc1-4fe0-b966-18f0badfae37
- Native: https://jam.dev/c/fc4862ad-590a-4a53-9675-92604f73a75e

### UNI-V2 Withdraw:

- SAFE: https://jam.dev/c/775bd1ad-a17a-4e30-ba9c-3850b23a7aac
- Native: https://jam.dev/c/84f7237d-42b0-4e60-9ca7-7f8f1af79209

### FOX Farming Deposit:

- Native: https://jam.dev/c/0a04501d-a62e-460b-94be-3ccf7a6e6f95
- SAFE: https://jam.dev/c/4727d4fc-94f6-42ce-9c28-2addea51fdfd

### FOX Farming Withdraw:

- SAFE: https://jam.dev/c/c2ee041a-a4ea-4527-b451-b36167a3442d
- Native: https://jam.dev/c/1f7e16ec-920b-409b-af99-695ce7d77518

Also fixed ETH/FOX staking claims the same way in 89925aaedb273ec86a81a1e92dcee178d561aa46 - tested in gome since borked in localhost:

- SAFE: https://jam.dev/c/78391a96-a5e3-4900-bf8f-b67fdccc00bd
- Native: https://jam.dev/c/0c26fd33-4be6-4d56-81d7-2aa0e8d48f1d

kaladinlight commented on packages/swapper/src/swappers/ArbitrumBridgeSwapper/endpoints.ts:
```diff
@@ -159,6 +158,9 @@ export const arbitrumBridgeApi: SwapperApi = {
     txHash,
     chainId,
     assertGetEvmChainAdapter,
+    getEthersV5Provider,
```
why are we passing this function vs importing the singleton directly?

gomesalexandre replied:
Because swapper deps 😢

kaladinlight replied:
easy to remove, but not necessary for this pr. threw me off since it was being added back from deps is all.

gomesalexandre replied:
9a73e2021fdd9b06045d067664dcc3b083fe5f1a

kaladinlight commented on packages/swapper/src/types.ts:
```diff
@@ -159,6 +160,8 @@ export type GetTradeQuoteInput =
 export type EvmSwapperDeps = {
   assertGetEvmChainAdapter: (chainId: ChainId) => EvmChainAdapter
+  fetchIsSmartContractAddressQuery: (userAddress: string, chainId: ChainId) => Promise<boolean>
+  getEthersV5Provider: (chainId: EvmChainId) => ethersV5.providers.JsonRpcProvider
```
does look like added back and not needed actually.

gomesalexandre replied:
Very nice catch! This guy now lives in contracts package so doesn't need to be dependency injected. 9a73e2021fdd9b06045d067664dcc3b083fe5f1a

kaladinlight commented on src/lib/getTxLink.ts:
```diff
@@ -33,7 +33,7 @@ type GetTxLink = GetTxBaseUrl &
   ({ txId: string; tradeId?: never } | { tradeId: string; txId?: never }) & {
     // TODO(gomes): make me required in a follow-up and have SAFE TxIds working holistically
```
applicable or still meant as follow up?

gomesalexandre replied:
Nice catch! Removed comment and made it required in 9dbfbe751b9b05d230ef09a09e0e517c4cb4f218

gomesalexandre commented on packages/swapper/src/safe-utils/index.ts:
```diff
@@ -10,10 +11,42 @@ import type {
 }: FetchSafeTransactionArgs): Promise<SafeTxInfo> => {
-  if (!isEvmChainId(chainId)) return { transaction: null, isSafeTxHash: false }
+  if (!accountId) {
+    return {
```
f0e188c2a3

0xApotheosis commented on src/features/defi/providers/fox-farming/components/FoxFarmingManager/Claim/ClaimStatus.tsx:
```diff
@@ -217,31 +275,21 @@ export const ClaimStatus: React.FC<ClaimStatusProps> = ({ accountId }) => {
-                  ? 'modals.status.estimatedGas'
-                  : 'modals.status.gasUsed',
-              )}
+              {translate(status ? 'modals.status.estimatedGas' : 'modals.status.gasUsed')}
```
Are we intentionally changing the logic here such that _any_ status will use the `modals.status.estimatedGas` translation?

gomesalexandre replied:
Nice catch - that wasn't intentional indeed! 82e3beffa5

0xApotheosis commented on src/features/defi/providers/fox-farming/components/FoxFarmingManager/Deposit/components/Status.tsx:
```diff
@@ -94,25 +102,100 @@ export const Status: React.FC<StatusProps> = ({ accountId }) => {
   const confirmedTransaction = useAppSelector(gs => selectTxById(gs, serializedTxIndex))
 
+  const { statusIcon, status, statusText, statusBg, statusBody } = useMemo(() => {
+    if (maybeSafeTx?.isQueuedSafeTx)
+      return {
+        statusIcon: null,
+        statusText: StatusTextEnum.pending,
+        status: TxStatus.Pending,
+        statusBody: translate('common.safeProposalQueued', {
+          currentConfirmations: maybeSafeTx?.transaction?.confirmations?.length,
```
There seems to be a lot of duplicate code surrounding the `Status` components, perhaps we can consolidate into helpers? Not for this PR of course, it's been open too long already!

gomesalexandre replied:
Added to my TODOs, happy to follow-up with that! 🙏🏽

0xApotheosis commented on src/pages/RFOX/components/Stake/StakeStatus.tsx:
```diff
@@ -51,9 +55,39 @@ export const StakeStatus: React.FC<StakeRouteProps & StakeStatusProps> = ({
   const bodyContent: BodyContent | null = useMemo(() => {
     if (!stakingAsset) return null
 
+    if (maybeSafeTx?.isQueuedSafeTx) {
+      return {
+        key: TxStatus.Pending,
+        title: [
```
A lot of duplicate code in these parts, too - again, not worth holding up this PR for, not blocking.

0xApotheosis commented on src/features/defi/providers/foxy/components/FoxyManager/Overview/Claim/ClaimStatus.tsx:
```diff
@@ -36,28 +38,28 @@ interface ClaimStatusState {
   chainId: ChainId
 }
 
-enum TxStatus {
-  PENDING = 'pending',
-  SUCCESS = 'success',
-  FAILED = 'failed',
-}
```
Oh man, thank you for this.

0xApotheosis commented on src/features/defi/providers/univ2/components/UniV2Manager/Deposit/components/Status.tsx:
```diff
@@ -44,6 +48,13 @@ export const Status: React.FC<StatusProps> = ({ accountId }) => {
+    accountId,
+  })
+
+  console.log({ maybeSafeTx })
```
rm?

gomesalexandre replied:
2319e87377

0xApotheosis reviewed (approved):
Overall I'm happy with this guy, what a monster.

There is likely opportunity to reduce duplication around a lot of the TX logic, but considering the size of this PR it doesn't seem sane to hold it up for that.

Did some basic smoke tests around TX status and explorer URLs and it's looking happy, will rely on ops when it hits release to cover the delta not already covered by Neo's excellent testing.

gomesalexandre commented:
As discussed with @0xApotheosis, going to merge and tackle comments as follow-up so this guy gets his own release 🙏🏽 tyvm  to yourself and @NeOMakinG for making landing this guy possible!
